src/backend/distributed/ddl/foreign_constraint.c

```diff
@@ -309,6 +369,94 @@ ErrorIfUnsupportedForeignConstraint(Relation relation, char distributionMethod,
 }
 
 
+bool
+ForeignKeyExistsToReferenceTableOnColumn(char *affectedColumnName,
+										 Oid relationId)
+{
+	Relation pgConstraint = NULL;
+	SysScanDesc scanDescriptor = NULL;
+	ScanKeyData scanKey[1];
+	int scanKeyCount = 1;
+	HeapTuple heapTuple = NULL;
+	bool foreignKeyToReferenceTableIncludesGivenColumn = false;
+	bool isNull = false;
+	Datum columnsDatum;
+	Datum *columnArray;
+	int columnCount = 0;
+	int attrIdx;
+
+	pgConstraint = heap_open(ConstraintRelationId, AccessShareLock);
+
+	ScanKeyInit(&scanKey[0],
+				Anum_pg_constraint_connamespace,
+				BTEqualStrategyNumber, F_OIDEQ,
+				get_rel_namespace(relationId));
+
+	scanDescriptor = systable_beginscan(pgConstraint, ConstraintNameNspIndexId, true,
+										NULL,
+										scanKeyCount, scanKey);
+
+	heapTuple = systable_getnext(scanDescriptor);
+	while (HeapTupleIsValid(heapTuple))
+	{
+		Oid referencedTableId;
+		Oid referencingTableId;
+		int pgConstraintKey = 0;
+		Form_pg_constraint constraintForm = (Form_pg_constraint) GETSTRUCT(heapTuple);
+
+		if (constraintForm->contype != CONSTRAINT_FOREIGN)
+		{
+			heapTuple = systable_getnext(scanDescriptor);
+			continue;
+		}
+
+		referencedTableId = constraintForm->confrelid;
+		referencingTableId = constraintForm->conrelid;
+
+		if (referencedTableId == relationId)
+		{
+			pgConstraintKey = Anum_pg_constraint_confkey;
+		}
+		else if (referencingTableId == relationId)
+		{
+			pgConstraintKey = Anum_pg_constraint_conkey;
+		}
+
+		columnsDatum = SysCacheGetAttr(CONSTROID, heapTuple,
+									   pgConstraintKey, &isNull);
+		deconstruct_array(DatumGetArrayTypeP(columnsDatum), INT2OID, 2, true,
+						  's', &columnArray, NULL, &columnCount);
```

Comment:
I think the calculation for `columnCount` is wrong. It's only the number of columns in the fkey. However, we should get the column count of the relation

Reply:
I don't agree. we should get the number of columns in the key because we use this count as a index while iterating on the columns in the key, not in the table. We get the columnArray from the same function call and use it in line 435.